### PR TITLE
FEATURE: Support inheritance from Fmi2Slave

### PR DESCRIPTION
**Pull request by Jorgelmh**
When building a Python FMU, the builder will try to find the class name that needs to be instantiated from the C++ wrapper by running a regex expression that aims to locate the class that extends Fmi2Slave in the python file specified when using the CLI. This regex implementation makes it impossible to have inheritance from the Fmi2Slave class to create more complex python models or libraries built on top of PythonFMU.

To fix this issue, we have introduced these changes.

- Update `get_class_name` in Builder.py to use the Python's Method Resolution Order (MRO) to identify which class has the deepest inheritance from the Fmi2Slave class.
- Similar solution but for the C++ wrapper, so that the CPPFMU that instantiates the Python class is able to find the right Python class.

### CODE REVIEW

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp [hunk 1]
```diff
@@ -150,7 +210,7 @@ void PySlaveInstance::initialize(PyGILState_STATE gilState)
     Py_XDECREF(pMessages_);
 
     PyObject* args = PyTuple_New(0);
-    PyObject* kwargs = Py_BuildValue("{ss,ss,sn,si}",
+    PyObject* kwargs = Py_BuildValue("{ss,ss,sL,si}",
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:213]: `sL` -> `sn` was changed recently. Is this intended or merge artifact?
See https://github.com/NTNU-IHB/PythonFMU/pull/181

**Comment by markaren** [comment]
One important question to settle here, is if inheritance is something that is needed / wanted. What does this enable that composition does not solve?

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp [hunk 2]
```diff
@@ -107,41 +169,39 @@ void PySlaveInstance::clearLogBuffer() const
...
+        } else {
+            const_cast<cppfmu::Logger&>(logger_).Log(statusValue, categoryStr, msgStr);
         }
-        PyList_SetSlice(pMessages_, 0, size, nullptr);
     }
     Py_DECREF(debugField);
     Py_DECREF(msgField);
     Py_DECREF(categoryField);
     Py_DECREF(statusField);
+    PyList_SetSlice(pMessages_, 0, size, nullptr);
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:204]: Bugfix or just re-arranging?

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp (hunk 1 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:213]: Probably a merge artifact, I'll revert!

File: pythonfmu/enums.py [hunk 3]
```diff
@@ -22,6 +22,7 @@ class Fmi2Initial(Enum):
     exact = 0
     approx = 1
     calculated = 2
+    none = 3 # additional value to allow for the cases when initial: --
```
Comment (markaren) [inline pythonfmu/enums.py:25]: Would be great if this was a separate PR

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp [hunk 4]
```diff
@@ -65,33 +143,17 @@ PySlaveInstance::PySlaveInstance(std::string instanceName, std::string resources
             handle_py_exception("[ctor] PyObject_GetAttrString", gilState);
         }
         int success = PyList_Insert(sys_path, 0, PyUnicode_FromString(resources_.c_str()));
+        PyList_Append(sys_path, PyUnicode_DecodeFSDefault(resources_.c_str()));
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:146]: This is a change unrelated to this feature request

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp [hunk 5]
```diff
@@ -65,33 +143,17 @@ PySlaveInstance::PySlaveInstance(std::string instanceName, std::string resources
...
-            handle_py_exception("[ctor] Py_BuildValue", gilState);
-        }
-
-        pClass_ = PyObject_GetAttr(pModule, pyClassName);
-        Py_DECREF(pyClassName);
-        Py_DECREF(pModule);
+        pClass_ = findClass(resources_, moduleName);
         if (pClass_ == nullptr) {
-            handle_py_exception("[ctor] PyObject_GetAttr", gilState);
+            handle_py_exception("[ctor] Py_BuildValue", gilState);
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:156]: Should perhaps be handled on call site. This way it's confusing

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp [hunk 6]
```diff
@@ -107,41 +169,39 @@ void PySlaveInstance::clearLogBuffer() const
     PyObject* categoryField = Py_BuildValue("s", "category");
     PyObject* statusField = Py_BuildValue("s", "status");
 
-    if ( pMessages_ != NULL && PyList_Check(pMessages_) ) {
-        auto size = PyList_Size(pMessages_);
-        for (auto i = 0; i < size; i++) {
-            PyObject* msg = PyList_GetItem(pMessages_, i);
+    auto size = PyList_Size(pMessages_);
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:172]: This is also a change unrelated to the feature proposed.

File: pythonfmu/pythonfmu-export/src/pythonfmu/PyState.hpp [hunk 7]
```diff
@@ -1,4 +1,3 @@
-
 #ifndef PYTHONFMU_PYTHONSTATE_HPP
 #define PYTHONFMU_PYTHONSTATE_HPP
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/PyState.hpp:3]: diff adds noise to the PR

File: pythonfmu/pythonfmu-export/src/pythonfmu/IPyState.hpp [hunk 8]
```diff
@@ -15,4 +15,4 @@ namespace pythonfmu
     };
 }
 
-#endif //PYTHONFMU_IPYTHONSTATE_HPP
```
Comment (markaren) [inline pythonfmu/pythonfmu-export/src/pythonfmu/IPyState.hpp:18]: diff adds noise to the PR

File: pythonfmu/enums.py (hunk 3 above)
Comment (Jorgelmh) [reply pythonfmu/enums.py]: Removed and will be tackled in a seconds PR

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp (hunk 4 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp]: Removed!

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp (hunk 5 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:156]: I don't understand what you mean by that, could you explain in more detail?

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp (hunk 6 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp]: Removed from PR

File: pythonfmu/pythonfmu-export/src/pythonfmu/IPyState.hpp (hunk 8 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/IPyState.hpp]: Removed from PR

File: pythonfmu/pythonfmu-export/src/pythonfmu/PyState.hpp (hunk 7 above)
Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PyState.hpp]: Removed

File: pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp (hunk 5 above)
Comment (markaren) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp:156]: I mean that `Py_BuildValue` is not called in this scope, but rather on line 107. Seeing this code at this point is confusing as context is missing. Changing the error to `handle_py_exception("[ctor] findClass", gilState);` might improve it (rather than moving to the function that provides the `nullptr`).

Comment (Jorgelmh) [reply pythonfmu/pythonfmu-export/src/pythonfmu/PySlaveInstance.cpp]: Solved! Thank you!
